#include "network.h"
#include "blockchain.h"
#include "transaction.h"
#include <json/json.h>
#include <iostream>
#include <thread>
#include <vector>
#include <sstream>
#include <array>
#include <unordered_set>
#include "syncing.h"
#include "generated/block_protos.pb.h"
#include "generated/transaction_protos.pb.h"
#include <filesystem>
#include <boost/asio.hpp>
#include <boost/asio/connect.hpp>
#include <boost/asio/ip/tcp.hpp>
#include <boost/asio/io_context.hpp>

#define ENABLE_DEBUG 0
namespace fs = std::filesystem;

// ‚úÖ Correct Constructor:
Network::Network(unsigned short port, Blockchain* blockchain)
    : port(port),
      blockchain(blockchain),
      isRunning(false),
      syncing(true),
      ioContext(),
      acceptor(ioContext, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), port)) 
{
    isRunning = true;
    listenerThread = std::thread(&Network::listenForConnections, this);
}

// ‚úÖ Correct Destructor:
Network::~Network() {
    try {
        ioContext.stop();
        acceptor.close();
        if (listenerThread.joinable()) {
            listenerThread.join();
        }
        std::cout << "‚úÖ Network instance cleaned up safely." << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "‚ùå Error during Network destruction: " << e.what() << std::endl;
    }
}
//
void Network::listenForConnections() {
    std::cout << "üåê Listening for connections on port: " << port << std::endl;
    while (isRunning) {
        std::shared_ptr<tcp::socket> socket = std::make_shared<tcp::socket>(ioContext);
        acceptor.accept(*socket);

        std::string senderIP = socket->remote_endpoint().address().to_string();
        std::string peerAddr = senderIP + ":" + std::to_string(socket->remote_endpoint().port());

        if (peerSockets.find(peerAddr) == peerSockets.end()) {
            peerSockets[peerAddr] = socket;
            savePeers();
        }

        std::cout << "üì° New peer connected: " << peerAddr << std::endl;
        std::thread(&Network::handlePeer, this, socket).detach();
    }
}

//

void Network::start() {
    startServer(); // Start listening for connections
    syncWithPeers(); // Automatically sync with peers at startup ‚úÖ
}

// ‚úÖ **Auto-Mining Background Thread**
void Network::autoMineBlock() {
    std::thread([this]() {
        while (true) {
            std::this_thread::sleep_for(std::chrono::seconds(5));

            Blockchain& blockchain = Blockchain::getInstance(8333);
            if (!blockchain.getPendingTransactions().empty()) {
                std::cout << "‚õèÔ∏è New transactions detected. Starting mining..." << std::endl;
                Block minedBlock = blockchain.minePendingTransactions("Miner");
                
                if (blockchain.isValidNewBlock(minedBlock)) {  // üîç Validate before adding
                    blockchain.addBlock(minedBlock);
                    Blockchain::getInstance().saveToDB();
                    broadcastBlock(minedBlock);
                    std::cout << "‚úÖ Mined & broadcasted block: " << minedBlock.getHash() << std::endl;
                }
            }
        }
    }).detach();
}

//
void Network::broadcastMessage(const std::string& message) {
    std::lock_guard<std::mutex> lock(peersMutex); // Thread safety
    for (const auto& peer : peerSockets) {
        sendMessageToPeer(peer.first, message);
    }
}

// ‚úÖ **Getter function for syncing status**
bool Network::isSyncing() const {
    return syncing;
}
// ‚úÖ **Connect to a peer and send a message**
void Network::sendMessage(std::shared_ptr<boost::asio::ip::tcp::socket> socket, const std::string& message) {
    try {
        if (!socket || !socket->is_open()) {
            return;
        }

        boost::asio::write(*socket, boost::asio::buffer(message + "\n"));
        std::cout << "üì° Sent message: " << message << std::endl;
    } catch (const std::exception& e) {
    }
}
//
void Network::sendMessageToPeer(const std::string& peer, const std::string& message) {
    try {
        boost::asio::io_context io_context;
        boost::asio::ip::tcp::resolver resolver(io_context);

        std::size_t colonPos = peer.find(":");
        if (colonPos == std::string::npos) {
            return;
        }

        std::string host = peer.substr(0, colonPos);
        std::string port = peer.substr(colonPos + 1);

        if (host.empty() || port.empty()) {
            return;
        }

        boost::asio::ip::tcp::resolver::results_type endpoints = resolver.resolve(host, port);
        boost::asio::ip::tcp::socket socket(io_context);
        boost::asio::connect(socket, endpoints);

        boost::asio::write(socket, boost::asio::buffer(message + "\n"));
        std::cout << "üì° Sent message to peer " << peer << ": " << message << std::endl;
    } catch (const std::exception& e) {
    }
}

// ‚úÖ **Broadcast a transaction to all peers**

void Network::broadcastTransaction(const Transaction& tx) {
    std::string txData = tx.serialize();

    for (const auto& peer : peerSockets) {
        auto socket = peer.second;
        if (socket && socket->is_open()) {
            try {
                boost::asio::write(*socket, boost::asio::buffer(txData + "\n"));
                std::cout << "üì° Transaction broadcasted to peer: " << peer.first << std::endl;
            } catch (const std::exception& e) {
                std::cerr << "‚ùå [ERROR] Failed to broadcast transaction to " << peer.first << ": " << e.what() << std::endl;
            }
        }
    }
}

// sync with peers
void Network::syncWithPeers() {
    std::cout << "üîÑ [INFO] Syncing with peers..." << std::endl;

    if (peerSockets.empty()) {
        std::cerr << "‚ö†Ô∏è [WARNING] No peers available for sync!\n";
        return;
    }

    Blockchain& blockchain = Blockchain::getInstance();  // ‚úÖ Use Singleton

    for (const auto& [peer, socket] : peerSockets) {
        if (peer.empty()) continue;

        std::cout << "üì° [DEBUG] Requesting blockchain sync from " << peer << "...\n";
        std::string response = requestBlockchainSync(peer);

        if (response.empty()) {
            std::cerr << "‚ùå [ERROR] No data received from " << peer << "!\n";
            continue;
        }

        // Debugging: Print received blockchain data
        std::cout << "üì° [DEBUG] Raw Blockchain Data from Peer (" << peer << "): "
                  << response.substr(0, 500) << "...\n";

        Blockchain& tempChain = Blockchain::getInstance();  // ‚úÖ Use Singleton

        if (!tempChain.deserializeBlockchain(response)) {
            std::cerr << "‚ùå [ERROR] Failed to parse Protobuf blockchain data from peer: " << peer << "\n";
            continue;
        }

        // ‚úÖ Merge only if received blockchain is longer & valid
        if (tempChain.getBlockCount() > blockchain.getBlockCount()) {
            std::cout << "‚úÖ [INFO] Longer blockchain detected! Merging...\n";
            blockchain.mergeWith(tempChain);
            blockchain.saveToDB();
        } else {
            std::cout << "‚úÖ [INFO] Local blockchain is up-to-date. No merge needed.\n";
        }
    }
}
//
void Network::connectToPeer(const std::string& ip, short port) {
    try {
        boost::asio::ip::tcp::resolver resolver(ioContext);
        auto endpoints = resolver.resolve(ip, std::to_string(port));

        auto socket = std::make_shared<boost::asio::ip::tcp::socket>(ioContext);
        boost::asio::connect(*socket, endpoints);

        {
            std::lock_guard<std::mutex> lock(peersMutex);
            peerSockets[ip + ":" + std::to_string(port)] = socket;
        }

        std::cout << "‚úÖ Connected to peer: " << ip << ":" << port << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "‚ùå Failed to connect to peer: " << ip << ":" << port << " | Error: " << e.what() << std::endl;
    }
}

// ‚úÖ **Broadcast peer list to all connected nodes**
void Network::broadcastPeerList() {
    if (peerSockets.empty()) return;

    Json::Value peerListJson;
    peerListJson["type"] = "peer_list";

    for (const auto& [peerAddr, _] : peerSockets) {
        if (peerAddr.find(":") == std::string::npos) {
            continue;
        }
        peerListJson["data"].append(peerAddr);
    }

    Json::StreamWriterBuilder writer;
    std::string peerListMessage = Json::writeString(writer, peerListJson);

    for (const auto& [peerAddr, _] : peerSockets) {
        sendData(peerAddr, peerListMessage);
    }
}

// ‚úÖ **Request peer list from connected nodes**
void Network::requestPeerList() {
for (const auto& [peerAddr, socket] : peerSockets) {
    sendData(peerAddr, R"({"type": "request_peers"})");

    }

    std::cout << "üì° Requesting peer list from all known peers..." << std::endl;
}
//
void Network::receiveFullChain(const std::string& senderIP, const std::string& data) {
    std::cout << "[INFO] Receiving full blockchain from " << senderIP << std::endl;

    std::lock_guard<std::mutex> lock(blockchainMutex);

    if (!Blockchain::getInstance().deserializeBlockchain(data)) {
        std::cerr << "[ERROR] Failed to deserialize blockchain data from " << senderIP << "!\n";
        return;
    }

    Blockchain::getInstance().saveToDB();
    std::cout << "[INFO] Blockchain successfully updated from " << senderIP << std::endl;
}
//network node
bool Network::connectToNode(const std::string& peerIP, int port) {
    try {
        if (peerIP == "127.0.0.1" && port == this->port) {
            std::cerr << "‚ö†Ô∏è Skipping self-connection attempt.\n";
            return false;  // Prevent connecting to own node
        }

        std::string fullPeer = peerIP;
        if (peerIP.find(":") == std::string::npos) {
            fullPeer += ":" + std::to_string(port);
        }

        if (peerSockets.find(fullPeer) != peerSockets.end()) {
            return false;
        }

        boost::asio::io_context ioContext;
        auto socketPtr = std::make_shared<boost::asio::ip::tcp::socket>(ioContext);
        boost::asio::ip::tcp::resolver resolver(ioContext);
        auto endpoints = resolver.resolve(peerIP, std::to_string(port));

        boost::asio::connect(*socketPtr, endpoints);

        peerSockets.emplace(fullPeer, std::move(socketPtr));

        std::cout << "‚úÖ Connected to new peer: " << fullPeer << std::endl;
        return true;
    } catch (std::exception& e) {
        std::cerr << "‚ùå Error connecting to node: " << e.what() << std::endl;
        return false;
    }
}

// ‚úÖ **Run Network Thread**
void Network::run() {
    serverThread = std::thread([this]() { startServer(); });
    serverThread.detach();  // ‚úÖ Keeps server running in background

    std::this_thread::sleep_for(std::chrono::seconds(2)); // ‚úÖ Allow some time for initialization

    requestPeerList();       // üîÑ Request peers on startup
    scanForPeers();          // üîÑ Try discovering new peers
    requestBlockchainSync(peerSockets.begin()->first);
    autoMineBlock();         // ‚õèÔ∏è Auto-mining thread
    periodicSync();          // üîÑ Keep syncing every few seconds
}
//
// ‚úÖ Auto-Discover Peers Instead of Manually Adding Nodes
std::vector<std::string> Network::discoverPeers() {
    std::vector<std::string> peers;

    std::ifstream file("data/peers.list");
    if (!file) {
        std::cerr << "‚ö†Ô∏è [WARNING] No known peers found. Bootstrap required!" << std::endl;
        return peers;
    }

    std::string peer;
    while (std::getline(file, peer)) {
        if (!peer.empty()) peers.push_back(peer);
    }

    file.close();
    return peers;
}

void Network::connectToDiscoveredPeers() {
    std::vector<std::string> peers = discoverPeers();
    for (const std::string& peer : peers) {
        connectToNode(peer, DEFAULT_PORT);
    }
}

//
void Network::periodicSync() {
    for (const auto& peer : peerSockets) {
        if (peer.first.empty()) continue;  // ‚úÖ Prevent empty peer errors

        std::cout << "üì° [DEBUG] Periodic sync request to " << peer.first << "\n";
        std::string response = requestBlockchainSync(peer.first);

        if (response.empty()) {
            std::cerr << "‚ö†Ô∏è [WARNING] Skipping peer " << peer.first << " due to empty response.\n";
            continue;
        }
    }
}

//
std::vector<std::string> Network::getPeers() {
    std::vector<std::string> peerList;
    for (const auto& peer : peerSockets) {
        peerList.push_back(peer.first);  // Extract peer addresses
    }
    return peerList;
}

// ‚úÖ **Handle Incoming Data with Protobuf Validation**
void Network::handleIncomingData(const std::string& senderIP, const std::string& data) {
    if (data.empty()) {
        std::cerr << "‚ùå [ERROR] Received empty data from peer: " << senderIP << "!\n";
        return;
    }

    std::cout << "üì° [DEBUG] Incoming Data (" << senderIP << ") First 500 Chars:\n"
              << data.substr(0, std::min(size_t(500), data.size())) << "...\n";

    // === Handle Blockchain Sync ===
    if (data.rfind("BLOCKCHAIN_DATA|", 0) == 0) {
        std::string blockchainData = data.substr(std::string("BLOCKCHAIN_DATA|").length());

        if (blockchainData.empty()) {
            std::cerr << "‚ùå [ERROR] Empty blockchain data received!\n";
            return;
        }

        alyncoin::BlockchainProto protoChain;
        if (!protoChain.ParseFromString(blockchainData)) {
            std::cerr << "‚ùå [ERROR] Failed parsing BlockchainProto!\n";
            return;
        }

        Blockchain& blockchain = Blockchain::getInstance();
        if (!blockchain.deserializeBlockchain(blockchainData)) {
            std::cerr << "‚ùå Blockchain deserialization failed!\n";
            return;
        }

        blockchain.saveToDB();
        std::cout << "‚úÖ Blockchain updated from " << senderIP << "\n";
        return;
    }

    // === Handle Block Data ===
    if (data.rfind("BLOCK_DATA|", 0) == 0) {
        std::string blockData = data.substr(std::string("BLOCK_DATA|").length());
        if (blockData.empty()) {
            std::cerr << "‚ùå [ERROR] Empty block data received!\n";
            return;
        }

        alyncoin::BlockProto protoBlock;
        if (!protoBlock.ParseFromString(blockData)) {
            std::cerr << "‚ùå [ERROR] BlockProto parsing failed!\n";
            return;
        }

        Block newBlock;
        if (!newBlock.deserializeFromProtobuf(protoBlock)) {
            std::cerr << "‚ùå [ERROR] Block deserialization failed!\n";
            return;
        }

        handleNewBlock(newBlock);
        return;
    }

    // === Handle Transaction Data ===
    try {
        Transaction tx = Transaction::deserialize(data);
        std::string senderPubKey = getPublicKeyPath(tx.getSender());

        if (tx.isValid(senderPubKey)) {
            blockchain->addTransaction(tx);
            blockchain->saveTransactionsToDB();
            std::cout << "‚úÖ Valid transaction received from " << senderIP << "\n";
        } else {
            std::cerr << "‚ùå Invalid transaction from " << senderIP << "\n";
        }
    } catch (const std::exception& e) {
        std::cerr << "‚ö†Ô∏è Exception while handling transaction from " << senderIP << ": " << e.what() << "\n";
    }
}

// ‚úÖ **Broadcast a mined block to all peers*
void Network::broadcastBlock(const Block& block) {
    alyncoin::BlockProto blockProto = block.toProtobuf();
    std::string serializedBlock;
    blockProto.SerializeToString(&serializedBlock);

    std::string message = "BLOCK_DATA|" + serializedBlock;
    for (const auto& peer : getPeers()) {
        sendData(peer, message);
    }
    std::cout << "üì° Broadcasted mined block to peers!\n";
}

//
void Network::receiveTransaction(const Transaction& tx) {
    std::string txHash = tx.getHash();
    if (seenTxHashes.count(txHash) > 0) return; // Already processed
    seenTxHashes.insert(txHash);

    Blockchain::getInstance().addTransaction(tx);
    broadcastTransaction(tx); // Re-broadcast to peers
}

// Valid peer
bool Network::validatePeer(const std::string& peer) {
    if (peer.find(":") == std::string::npos) {  // ‚úÖ Correct format check
        return false;
    }

    if (peerSockets.find(peer) != peerSockets.end()) {
        return false;
    }

    std::string handshakeMessage = "PEER_HANDSHAKE";
    sendData(peer, handshakeMessage);

    std::cout << "‚úÖ Peer validated: " << peer << std::endl;
    return true;
}

//Handle new block
void Network::handleNewBlock(const Block& newBlock) {
    Blockchain& blockchain = Blockchain::getInstance();

    if (!newBlock.hasValidProofOfWork()) {
        std::cerr << "‚ùå [ERROR] Received block fails Proof-of-Work check! Rejecting...\n";
        return;
    }

    if (!blockchain.isValidNewBlock(newBlock)) {
        std::cerr << "‚ùå Invalid block received! Rejecting.\n";
        return;
    }

    std::string minerPublicKey = KEY_DIR + newBlock.getMinerAddress() + "_public.pem";
    if (!fs::exists(minerPublicKey)) {
        std::cerr << "‚ö†Ô∏è [WARNING] Miner public key missing. Generating...\n";
        Crypto::generateKeysForUser(newBlock.getMinerAddress());
        std::this_thread::sleep_for(std::chrono::milliseconds(500));

        if (!fs::exists(minerPublicKey)) {
            std::cerr << "‚ùå [ERROR] Miner public key generation failed!\n";
            return;
        }
    }

    if (!Crypto::verifyMessage(minerPublicKey, newBlock.getBlockSignature(), newBlock.getHash())) {
        std::cerr << "‚ùå Block signature verification failed! Rejecting block.\n";
        return;
    }

    blockchain.addBlock(newBlock);
    blockchain.saveToDB();
    std::cout << "‚úÖ Block successfully added to blockchain!\n";
}

// Black list peer
void Network::blacklistPeer(const std::string& peer) {
    peerSockets.erase(peer);
    bannedPeers.insert(peer);
}

bool Network::isBlacklisted(const std::string& peer) {
    return bannedPeers.find(peer) != bannedPeers.end();
}

// ‚úÖ **Send Data to Peer with Error Handling**
bool Network::sendData(const std::string& peer, const std::string& data) {
    auto it = peerSockets.find(peer);
    if (it == peerSockets.end() || !it->second || !it->second->is_open()) {
        std::cerr << "‚ùå [ERROR] Peer socket not found or not open: " << peer << "\n";
        return false;
    }

    try {
        std::cout << "üì° [DEBUG] Sending message to " << peer << ": " << data << "\n";
        boost::asio::write(*it->second, boost::asio::buffer(data + "\n"));
        return true;
    } catch (const std::exception& e) {
        std::cerr << "‚ùå [ERROR] Failed to send data to " << peer << ": " << e.what() << "\n";
        return false;
    }
}

// ‚úÖ **Request Blockchain Sync from Peers**
std::string Network::requestBlockchainSync(const std::string& peer) {
    if (peerSockets.find(peer) == peerSockets.end()) {
        std::cerr << "‚ùå [ERROR] Peer not found: " << peer << "\n";
        return "";
    }

    std::cout << "üì° Requesting blockchain sync from: " << peer << "\n";

    // Prefix remains consistent
    if (!sendData(peer, "REQUEST_BLOCKCHAIN")) {
        std::cerr << "‚ùå Failed to send sync request to " << peer << "\n";
        return "";
    }

    // Fire-and-forget: we'll process received blockchain when the peer responds
    return "Sync request sent to " + peer;
}

// ‚úÖ **Start Listening for Incoming Connections**
void Network::startServer() {
    try {
        std::cout << "üåê Node is now listening for connections on port: " << port << "\n";
        acceptConnections();
        ioContext.run();
    } catch (const std::exception& e) {
        std::cerr << "‚ùå [ERROR] Server failed to start: " << e.what() << "\n";
        std::cerr << "‚ö†Ô∏è Try using a different port or checking if another instance is running.\n";
    }
}

// ‚úÖ **Receive Data from Peer (Blocking)**
std::string Network::receiveData(const std::string& peer) {
    try {
        auto it = peerSockets.find(peer);
        if (it == peerSockets.end()) {
            std::cerr << "‚ùå [ERROR] Peer not found: " << peer << std::endl;
            return "";
        }

        std::vector<char> buffer(1024);
        boost::system::error_code error;
        size_t length = it->second->read_some(boost::asio::buffer(buffer), error);

        if (error && error != boost::asio::error::eof) {
            std::cerr << "‚ùå [ERROR] Error reading data from " << peer << ": " << error.message() << std::endl;
            return "";
        }

        std::string receivedData(buffer.data(), length);
        std::cout << "üì• [DEBUG] Received Data from " << peer << ": " 
                  << receivedData.substr(0, 200) << "...\n";
        return receivedData;
    } catch (const std::exception& e) {
        std::cerr << "‚ùå [EXCEPTION] Exception in receiveData: " << e.what() << std::endl;
        return "";
    }
}

// ‚úÖ Accept connections
void Network::acceptConnections() {
    while (true) {
        auto socket = std::make_shared<boost::asio::ip::tcp::socket>(ioContext);
        acceptor.accept(*socket);

        std::string senderIP = socket->remote_endpoint().address().to_string();
        std::string peerAddr = senderIP + ":" + std::to_string(socket->remote_endpoint().port());

        {
            std::lock_guard<std::mutex> lock(peersMutex);
            if (peerSockets.find(peerAddr) == peerSockets.end()) {
                peerSockets[peerAddr] = socket;
                savePeers();
            }
        }

        std::cout << "üì° New peer connected: " << peerAddr << std::endl;
        std::thread(&Network::handlePeer, this, socket).detach();
    }
}

// ‚úÖ Add peer
void Network::addPeer(const std::string& peer) {
    if (peerSockets.find(peer) != peerSockets.end()) {
        return;
    }

    std::shared_ptr<boost::asio::ip::tcp::socket> socket = std::make_shared<boost::asio::ip::tcp::socket>(ioContext);
    peerSockets.emplace(peer, socket);

    std::cout << "üì° Peer added: " << peer << std::endl;
    savePeers();  // ‚úÖ Save immediately
}

// ‚úÖ Placeholder for handlePeer (implement logic as needed)
void Network::handlePeer(std::shared_ptr<tcp::socket> socket) {
    try {
        char buffer[1024];
        boost::system::error_code error;
        std::string peerId = socket->remote_endpoint().address().to_string();

        while (isRunning) {
            size_t bytesRead = socket->read_some(boost::asio::buffer(buffer), error);
            if (error == boost::asio::error::eof || bytesRead == 0) {
                std::cout << "Peer disconnected: " << peerId << "\n";
                break;
            }

            std::string data(buffer, bytesRead);
            handleIncomingData(peerId, data);
        }

        std::lock_guard<std::mutex> lock(peersMutex);
        peerSockets.erase(peerId);
        std::cout << "üîå Cleaned up peer socket: " << peerId << "\n";

    } catch (std::exception& e) {
        std::cerr << "‚ö†Ô∏è Exception in handlePeer: " << e.what() << "\n";
    }
}

//
void Network::sendLatestBlockIndex(const std::string& peerIP) {
    Json::Value msg;
    msg["type"] = "latest_block_index";
    msg["data"] = Blockchain::getInstance().getLatestBlock().getIndex();
    Json::StreamWriterBuilder writer;
    sendData(peerIP, Json::writeString(writer, msg));
}

void Network::handleReceivedBlockIndex(const std::string& peerIP, int peerBlockIndex) {
    int localIndex = Blockchain::getInstance().getLatestBlock().getIndex();
    if (peerBlockIndex > localIndex) {
        std::cout << "üì° Peer " << peerIP << " has longer chain. Requesting sync...\n";
        requestBlockchainSync(peerIP);
    } else {
        std::cout << "‚úÖ Local chain is up-to-date. No sync needed.\n";
    }
}

// ‚úÖ **Fix Peer Saving & Loading**
void Network::loadPeers() {
    std::ifstream file("peers.txt");
    if (!file.is_open()) {
        std::cerr << "‚ö†Ô∏è [WARNING] peers.txt not found. Attempting auto-discovery...\n";
        scanForPeers();  
        return;
    }

    std::string line;
    while (std::getline(file, line)) {
        if (line.empty() || line.find(":") == std::string::npos) continue;

        // ‚úÖ Filter self-peer
        if (line == "127.0.0.1:" + std::to_string(port)) {
            std::cerr << "‚ö†Ô∏è Skipping self-peer: " << line << "\n";
            continue;
        }

        peerSockets.insert({line, nullptr});
        std::cout << "‚úÖ Peer loaded & connected: " << line << "\n";

        // Auto-connect attempt
        if (!connectToNode(line.substr(0, line.find(":")), stoi(line.substr(line.find(":") + 1)))) {
            std::cerr << "‚ö†Ô∏è Failed to connect to loaded peer: " << line << "\n";
        }
    }

    file.close();
    std::cout << "‚úÖ Peers loaded and connected successfully!\n";
}

//
void Network::scanForPeers() {
    std::vector<std::string> potentialPeers = {
        "127.0.0.1:8080",
        "127.0.0.1:8333",
        "192.168.1.1:8080",
        "192.168.1.2:8333"
    };

    std::cout << "üîç Scanning for active AlynCoin nodes..." << std::endl;
    
    boost::asio::io_context ioContext;  // Declare io_context

    for (const auto& peer : potentialPeers) {
        if (connectToNode(peer.substr(0, peer.find(":")), std::stoi(peer.substr(peer.find(":") + 1)))) {
            std::cout << "‚úÖ Found & connected to: " << peer << std::endl;
            peerSockets[peer] = std::make_shared<boost::asio::ip::tcp::socket>(ioContext);  // ‚úÖ Corrected
            savePeers();
        }
    }

    if (peerSockets.empty()) {
        std::cout << "‚ö†Ô∏è No active peers found. Will retry periodically." << std::endl;
    }
}

// ‚úÖ **Ensure Peers are Saved Correctly**
void Network::savePeers() {
    std::ofstream file("peers.txt", std::ios::trunc);
    if (!file) {
        std::cerr << "‚ùå Error: Unable to save peers.txt!" << std::endl;
        return;
    }

    for (const auto& [peer, _] : peerSockets) {
        if (!peer.empty() && peer.find(":") != std::string::npos) {
            file << peer << std::endl;
        }
    }
    file.close();
    std::cout << "‚úÖ Peer list saved successfully." << std::endl;
}
// ‚úÖ **Send Latest Block to Peer (Optimized)**
void Network::sendLatestBlock(const std::string& peerIP) {
    Blockchain& blockchain = Blockchain::getInstance();

    if (blockchain.getChain().empty()) {
        std::cerr << "‚ö†Ô∏è Warning: Blockchain is empty! No block to send.\n";
        return;
    }

    Block latestBlock = blockchain.getLatestBlock();
    
    alyncoin::BlockProto protoBlock;
    latestBlock.serializeToProtobuf(protoBlock);  // ‚úÖ Convert to Protobuf

    std::string serializedBlock;
    protoBlock.SerializeToString(&serializedBlock);  // ‚úÖ Serialize to string

    sendData(peerIP, "BLOCK_DATA|" + serializedBlock);  // ‚úÖ Send Protobuf data
    std::cout << "üì° Sent latest block to " << peerIP << std::endl;
}

// ‚úÖ **Send Full Blockchain to Peer (Binary Storage)**
void Network::sendFullChain(const std::string& peer) {
    std::cout << "[INFO] Sending full blockchain to " << peer << std::endl;

    std::string serializedData;
    if (!Blockchain::getInstance().serializeBlockchain(serializedData)) {
        std::cerr << "[ERROR] Failed to serialize blockchain data!\n";
        return;
    }

    sendData(peer, serializedData);
}
// ‚úÖ **Process Incoming Data (Optimized)**
void Network::processReceivedData(const std::string& peer, const std::string& data) {
    Json::CharReaderBuilder reader;
    Json::Value jsonData;
    std::string errors;
    std::istringstream s(data);
    if (!Json::parseFromStream(reader, s, &jsonData, &errors)) return;

    std::string type = jsonData["type"].asString();
    
    if (type == "block") {
        alyncoin::BlockProto protoBlock;
        protoBlock.ParseFromString(jsonData["data"].asString()); // ‚úÖ Parse Protobuf

        Block newBlock;
        if (newBlock.deserializeFromProtobuf(protoBlock)) {
            Blockchain::getInstance().addBlock(newBlock);
        }
    } else if (type == "transaction") {
        alyncoin::TransactionProto protoTx;
        protoTx.ParseFromString(jsonData["data"].asString()); // ‚úÖ Parse Protobuf

        Transaction tx;
        tx.deserializeFromProtobuf(protoTx);
        Blockchain::getInstance().addTransaction(tx);
    }
}
// cleanup
void Network::cleanupPeers() {
    std::lock_guard<std::mutex> lock(peersMutex); // Always lock before iterating peerSockets
    std::vector<std::string> inactivePeers;

    for (const auto& peer : peerSockets) {
        try {
            if (!peer.second || !peer.second->is_open()) {
                std::cerr << "‚ö†Ô∏è Peer socket closed: " << peer.first << "\n";
                inactivePeers.push_back(peer.first);
                continue;
            }

            // Optional: send a ping
            boost::system::error_code ec;
            peer.second->send(boost::asio::buffer("PING"), 0, ec);
            if (ec) {
                std::cerr << "‚ö†Ô∏è Failed to ping peer: " << peer.first << " - Marking as inactive.\n";
                inactivePeers.push_back(peer.first);
            } else {
                std::cout << "‚úÖ Peer active: " << peer.first << "\n";
            }
        } catch (const std::exception& e) {
            std::cerr << "‚ö†Ô∏è Exception checking peer " << peer.first << ": " << e.what() << "\n";
            inactivePeers.push_back(peer.first);
        }
    }

    // Remove inactive peers
    for (const auto& peer : inactivePeers) {
        peerSockets.erase(peer);
        std::cout << "üóëÔ∏è Removed inactive peer: " << peer << "\n";
    }
}
